ChopChop/Models/Ingredients/Ingredient.swift

```diff
@@ -0,0 +1,271 @@
+                usedBatches.append(batch)
+            } catch QuantityError.negativeQuantity {
+                try batch.subtract(subtractedQuantity)
+                subtractedQuantity.value = 0
+                break
+            }
+        }
+
+        guard subtractedQuantity.value == 0 else {
+            throw IngredientError.insufficientQuantity
+        }
+
+        removeBatches(usedBatches)
+    }
+
+    /**
+     Combines this ingredient with another of the same name and quantity type.
+     Batches with the same expiry date are merged.
+     - Throws:
+        - `QuantityError.differentQuantityTypes`:
+            if the given quantity does not have the same type as the ingredient.
+        - `QuantityError.negativeQuantity`:
+            f any addition result is negative.
+        - `IngredientError.differentIngredients`:
+            if the given ingredient does not have the same name as this ingredient.
+     */
+    func combine(with ingredient: Ingredient) throws {
+        guard self.name == ingredient.name else {
+            throw IngredientError.differentIngredients
+        }
+
+        guard self.quantityType == ingredient.quantityType else {
+            throw QuantityError.differentTypes
+        }
+
+        for batch in ingredient.batches {
+            let quantity = batch.quantity
+            let expiryDate = batch.expiryDate
+            try add(quantity: quantity, expiryDate: expiryDate)
+        }
+    }
+}
+
+// MARK: - Batch operations
+extension Ingredient {
+    /**
+     Returns the batch with the given expiry date, or `nil` if it does not exist.
+     */
+    func getBatch(expiryDate: Date?) -> IngredientBatch? {
+        batches.first { batch in
+            batch.expiryDate == expiryDate
+        }
+    }
+
+    /**
+     Removes the batch identified by the given expiry date.
+     If such a batch does not exist, do nothing.
+     */
+    func removeBatch(expiryDate: Date?) {
+        batches.removeAll { batch in
+            batch.expiryDate == expiryDate
+        }
+    }
+
+    /**
+     Removes all batches with expiry dates earlier than the current date.
+     */
+    func removeExpiredBatches() {
+        let currentDate = Date()
+        var removedBatches: [IngredientBatch] = []
+
+        for batch in batches {
+            guard let expiryDate = batch.expiryDate else {
+                continue
+            }
+
+            if expiryDate < currentDate {
+                removedBatches.append(batch)
+            }
+        }
+
+        removeBatches(removedBatches)
+    }
+
+    private func removeBatches(_ removedBatches: [IngredientBatch]) {
+        for batch in removedBatches {
+            removeBatch(expiryDate: batch.expiryDate)
+        }
+    }
+}
+
+extension Ingredient: Equatable {
+    static func == (lhs: Ingredient, rhs: Ingredient) -> Bool {
+        lhs.quantityType == rhs.quantityType && lhs.name == rhs.name
+    }
+}
+
+enum IngredientError: Error {
```

Comment:
i'll follow your way of putting the enum errors in the file itself

Reply:
Should we discuss if having it in a separate file is better? I'm open to both ways

Reply:
I think we can leave it like this for now, I don't think one way is better than the other